plugins/overwatch.rb

```diff
@@ -0,0 +1,660 @@
+    #
+    # @return [void] A useful return value is not expected here
+    def session_log(sid, msg)
+      return unless self.config[:log]
+      ::File.open(::File.join(self.config[:base], "session.log"), "a") do |fd|
+        fd.puts "#{Time.now.strftime('%Y-%m-%d %H:%M:%S')} Session #{sid} [#{self.state[sid][:info]}] #{msg}"
+      end
+    end
+
+    ##################################################################################
+    # Route Management Functions
+    ##################################################################################
+
+    # Perform the routing operations
+    # Response_timeout is changed to the defined stale time to stay consistant
+    # and not wait 300 seconds for timeout.
+    #
+    # @sid [int class] Session ID of of the session currently being processed.
+    #
+    # @return [int array] Array of session id's for stale sessions based on routing settings.
+    def process_routing(sid)
+      return unless self.config[:autoroute]
+      return if self.state[sid][:error_count] >= 3
+      
+      if self.config[:reroute_stale]
+        stale_sids = get_stale_route_sids
+        return if stale_sids.include? sid # Skip this session if it has been tagged as routing stale
+      end
+
+      if framework.sessions[sid].respond_to?(:response_timeout)
+        last_known_timeout = framework.sessions[sid].response_timeout
+        framework.sessions[sid].response_timeout = self.config[:route_timeout] # Don't want to wait 300 secs for a time out!
+      end
+
+      add_routes(sid, stale_sids)
+
+      if framework.sessions[sid].respond_to?(:response_timeout) && last_known_timeout 
+        framework.sessions[sid].response_timeout = last_known_timeout # Set it back now that we are done trying to add routes.
+      end
+    end
+
+    # Step through all of the current sessions and look at their age.
+    # Return an array of sessions that are passed the defined timeout age (route_timeout).
+    # Skips sessions that are not compatabile with this check.
+    #
+    # @return [int array] Array of session id's for stale sessions based on routing settings.
+    def get_stale_route_sids
+      stale_sids = []
+      framework.sessions.keys.each do | sid |
+        next unless stale_check_compatible?(sid)
+        session_age = Time.now.to_i - framework.sessions[sid].last_checkin.to_i
+        stale_sids << sid if session_age >= self.config[:route_timeout]
+      end
+      return stale_sids
+    end
+
+    # Search for valid subnets on the target and attempt
+    # add a route to each.
+    #
+    # @sid [int class] Session ID of the current session
+    # @stale_sids [int array] Array of session id's of sessions marked as stale
+    #
+    # @return [void] A useful return value is not expected here
+    def add_routes(sid, stale_sids)
+      return unless route_compatible?(sid)
+      return unless framework.sessions[sid].alive?
+
+      switch_board = Rex::Socket::SwitchBoard.instance
```

Comment:
Is the reference instantiation required here, or would the namespace reference suffice for the :route_exists? call below? While the method-local reference might somewhat clean up the code, it creates another object for the GC to track which is often called from other contexts in the framework. Its a nitpick, take it or leave it. :)